staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go

```diff
@@ -59,8 +61,18 @@ func doTransformObject(ctx context.Context, obj runtime.Object, opts interface{}
 	if _, ok := obj.(*metav1.Status); ok {
 		return obj, nil
 	}
-	if err := setObjectSelfLink(ctx, obj, req, scope.Namer); err != nil {
-		return nil, err
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.RemoveSelfLink) {
+		if err := setObjectSelfLink(ctx, obj, req, scope.Namer); err != nil {
+			return nil, err
+		}
+	}
+
+	// Ensure that for empty lists we don't return <nil> items.
+	if meta.IsListType(obj) && meta.LenList(obj) == 0 {
+		if err := meta.SetList(obj, []runtime.Object{}); err != nil {
+			return nil, err
+		}
 	}
```

Comment:
off the top of my head, I can't think of a reason it would matter, but since SetList was conditionally done at the top of `setObjectSelfLink`, should we do that *before* calling `setObjectSelfLink` to be perfectly equivalent?